Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/AdapterBuilder.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzFramework/DocumentPropertyEditor/AdapterBuilder.h>
+#include <AzFramework/DocumentPropertyEditor/DocumentSchema.h>
+
+namespace AZ::DocumentPropertyEditor
+{
+    AdapterBuilder::AdapterBuilder(RoutingAdapter* routingAdapter)
+        : m_routingAdapter(routingAdapter)
+    {
+    }
+
+    void AdapterBuilder::Value(Dom::Value value)
+    {
+        CurrentNode().SetNodeValue(value);
+    }
+
+    void AdapterBuilder::Attribute(Name attribute, Dom::Value value)
+    {
+        CurrentNode()[attribute] = AZStd::move(value);
+    }
+
+    bool AdapterBuilder::IsError() const
+    {
+        return !m_error.empty();
+    }
+
+    const AZStd::string& AdapterBuilder::GetError() const
+    {
+        return m_error;
+    }
+
+    Dom::Value&& AdapterBuilder::FinishAndTakeResult()
+    {
+        AZ_Assert(m_entries.empty(), "AdapterBuilder::FinishAndTakeResult called before the builder finished");
```

Comment:
This should probably be a warning since I can see use cases for users wanting an early-out mechanism to get the dom built until it hits an error

Reply:
Hmm, this is more in "programmer error" territory if this is ever unbalanced, but I'll keep an eye on how this gets used and what's useful for error reporting.